### DISCUSSION TITLE
error: too many open files in system, (CAN NOT OPEN FILE)

### DISCUSSION THREAD

**Discussion by shelly-ctrl** [ClickHouse/ClickHouse#60891] [Q&A] [answered]
The clickhouse cluster architecture has 6 shards and two copies. The local table dwa_oym_cust_all_dm_load_local has 4,000 columns. There are 1,000 files every day. One file is about 400MB, and there are about 40,000 pieces of data in one file. The business is currently connected to one of the nodes when writing data. The business logic is that the data of these 1000 files are first sequentially written into the distributed total table dwa_oym_cust_all_dm_load_all. After all 1000 files are written, this distributed total table (dwa_oym_cust_all_dm_load_all) is queried and inserted into another distributed total table dwa_ovm_cust_all_dm_all. However, after running for a period of time (insert into dwa_ovm_cust_all_dm_all select * from dwa_oym_cust_all_dm_load_all), the error log reported an error of too many open files in system. The limit on the number of open files in the clickhouse.service service has been set to unlimited.

``` 
2024.03.01 16:50:15 .344919[ 393249 1(e756d658-87bc-449c-8a1c-72f37895539d} <Error>executeQuery: Code: 76. DB::BrrnoException: Cannot open file /data/clickhouse/clickhouse/data/ store/7db/7dbb284f-3d71-48fd-afa4-793bleldbdd0/tmp insert 20240228 1616l 0/partition.dat, errno: 23. str error: too many open files in system, (CAN NoT OPEN FILE) (version 23 6.1.1524 ( official buld)) (from [::ffff:10.243,238.217]:42926)(in query: insert into dwa_ovm_cust_all_dm_all select * from dwa_oym_cust_all_dm_load_all;),Stack trace (when copying this message, always include the lines below): 
```
```
# cat /etc/systemd/system/clickhouse-server.service
[Unit]
Description=ClickHouse Server (analytic DBMS for big data)
Requires=network-online.target
After=network-online.target

[Service]
Type=simple
User=clickhouse
Group=clickhouse
Restart=always
RestartSec=30
RuntimeDirectory=clickhouse-server
ExecStart=/usr/bin/clickhouse-server --config=/etc/clickhouse-server/config.xml --pid-file=/data/clickhouse/data/clickhouse-server.pid
LimitCORE=infinity
LimitNOFILE=infinity
CapabilityBoundingSet=CAP_NET_ADMIN CAP_IPC_LOCK CAP_SYS_NICE

[Install]
WantedBy=multi-user.target
```

The following parameters have been optimized according to the recommendations of the official website (https://clickhouse.com/docs/knowledgebase/Insert_select_settings_tuning):
```
<max_block_size>1048576000</max_block_size>
<min_insert_block_size_rows>1000000000</min_insert_block_size_rows>
<max_insert_block_size>1000000000</max_insert_block_size>
<min_insert_block_size_bytes>500000000</min_insert_block_size_bytes>
<parts_to_throw_insert>10000</parts_to_throw_insert> 
```
However, an error is still reported.

Help: The total amount of data is not very large. Why does insert into dwa_ovm_cust_all_dm_all select * from dwa_oym_cust_all_dm_load_all report an error of too many open files? Are too many temporary files generated? from a distributed
What is the detailed internal operating mechanism of the master table dwa_oym_cust_all_dm_load_all inserted into another distributed master table?

Thanks.

**Comment by shelly-ctrl** [comment]
The table structure is as follows:
[dwa_ovm_cust_all_dm_load_local.pdf](https://github.com/ClickHouse/ClickHouse/files/14505155/dwa_ovm_cust_all_dm_load_local.pdf)
[dwa_ovm_cust_all_dm_local.pdf](https://github.com/ClickHouse/ClickHouse/files/14505158/table.dwa_ovm_cust_all_dm_local.pdf)

**Comment by HwiLu** [comment, marked as answer]
I encountered the same problem, and my table had 10,000 columns. I later solved the problem by increasing the number of open files.

You can query the `system.query_log`   to see how many files are opened for each your query, and monitor the number of open files in Linux to see how many files it reaches before it starts to report an error.

**Reply by shelly-ctrl** [reply]
Thanks
